### PR TITLE
fixes #202 by retrying on an UNKNOWN error code

### LINKED ISSUE #202: Error: 2 UNKNOWN: Transaction outcome unknown
#### Environment details

  - OS: CoreOS
  - Node.js version: v8.9.4
  - npm version: v6.0.1
  - `@google-cloud/spanner` version: v1.4.1

#### Steps to reproduce

  1. ???
  2. ???

Not sure what causes this but these errors started trickling in this morning.  There's been about ~75 throughout the course of the day.  Mainly wondering what would cause this sort of error and how I can go about debugging it.

Error info:
```
 meta: {
   code: 2    
   details: "Transaction outcome unknown."    
   message: "2 UNKNOWN: Transaction outcome unknown."    
   metadata: {
      _internal_repr: {
    }
 }
 note: "Exception occurred in retry method that was not classified as transient"    
 stack: [
    0: "Error: 2 UNKNOWN: Transaction outcome unknown."     
    1: "at Object.exports.createStatusError (/opt/app/node_modules/google-gax/node_modules/grpc/src/common.js:87:15)"     
    2: "at Object.onReceiveStatus (/opt/app/node_modules/google-gax/node_modules/grpc/src/client_interceptors.js:1214:28)"     
    3: "at InterceptingListener._callNext (/opt/app/node_modules/google-gax/node_modules/grpc/src/client_interceptors.js:590:42)"     
    4: "at InterceptingListener.onReceiveStatus (/opt/app/node_modules/google-gax/node_modules/grpc/src/client_interceptors.js:640:8)"     
    5: "at callback (/opt/app/node_modules/google-gax/node_modules/grpc/src/client_interceptors.js:867:24)"     
    6: ""     
  ]
  statusCode: 500    
} 
```

### PR DESCRIPTION
**Pull request by crwilcox**
Fixes #202

- [x] Tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

**Comment by JustinBeckwith** [comment]
A unit test for this would be lovely :)

**Comment by crwilcox** [comment]
@JustinBeckwith , sure I thought about duplicating the one we have for errcode 10 so easy to add :)

File: src/transaction.js [hunk 1]
```diff
@@ -772,20 +779,30 @@ Transaction.prototype.runStream = function(query) {
...
   );
 };
 
+/**
+ * Specifies whether a specific error code can be retried.
+ *
+ * @param {number} errCode - the error code
+ * @return {boolean}
+ */
+Transaction.prototype.isRetryableErrorCode_ = function(errCode) {
```
Comment (stephenplusplus) [inline src/transaction.js]: Can you add a `describe` test block for `isRetryableErrorCode_`?

- it should return true for an aborted error code
- it should return true for an unknown error code

File: src/transaction.js [hunk 2]
```diff
@@ -772,20 +779,30 @@ Transaction.prototype.runStream = function(query) {
...
- * of an ABORTED error.
+ * of a retryable error.
  *
  * @param {error} err - The request error.
  * @return {boolean}
  */
 Transaction.prototype.shouldRetry_ = function(err) {
   return (
-    err.code === ABORTED &&
+    this.isRetryableErrorCode_(err.code) &&
```
Comment (stephenplusplus) [inline src/transaction.js]: We need a test in the `shouldRetry_` describe block that checks that `err.code` is passed to `this.isRetryableErrorCode_`.

File: test/transaction.js [hunk 3]
```diff
@@ -725,7 +725,39 @@ describe('Transaction', function() {
         fakeStream.emit('error', error);
       });
 
-      it('should retry the transaction', function(done) {
+      it('should retry the transaction for UNKNOWN', function(done) {
```
Comment (stephenplusplus) [inline test/transaction.js]: This test needs to mock `isRetryableErrorCode_` and make sure it receives the error code.

File: src/transaction.js [hunk 4]
```diff
@@ -403,7 +410,7 @@ Transaction.prototype.request = function(config, callback) {
   );
 
   this.session.request(config, function(err, resp) {
-    if (!self.runFn_ || !err || err.code !== ABORTED) {
+    if (!self.runFn_ || !err || !self.isRetryableErrorCode_(err.code)) {
```
Comment (stephenplusplus) [inline src/transaction.js]: We need a test in the `request` describe block that checks that `err.code` is passed to `this.isRetryableErrorCode_`.

File: test/transaction.js (hunk 3 above)
Comment (crwilcox) [reply test/transaction.js]: This test mocks out shouldRetry and shouldRetry now has a test for itself to verify it handles this. So this particular thing seems like testing in the wrong place to me.

Comment (crwilcox) [reply test/transaction.js]: I just saw why this is desired and have added this unit test.

**Pull request merged**